### ISSUE TITLE
Idea: Block based collator

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot-sdk#4813]
The currently getting implemented [slot based collator](https://github.com/paritytech/polkadot-sdk/issues/3168) is fully time based, but still depends on relay chain blocks to validate the time, to be able to prove legitimate aura block authorship.

The result is, that while we build based on time/parachain slots we still need to wait for the relay block for the current slot to arrive, before building a block.

Slots:

|i-----------|--------i---|-------i----|

In the above graphic, arrival of relay chain blocks is marked with an `i`.  The collator still needs to wait for that relay chain block in its slot, because the timestamp of the relay chain block is used to validate the "correctness" of the slot: The aura authorship is compared with the timestamp in the relay chain block: It has to be in the right slot for it to be valid.

Now, why is this a problem? This is actually only problematic if you are having a parachain with really high load, think of elastic scaling. Here the waiting for the relay chain block can significantly harm performance, we will simply be lacking the time to build blocks fast enough to occupy all our cores: One second waiting is one second not building.

## Towards a Solution

First idea to fix this was to use the relay chain block of the previous slot to validate the time. This approach has two problems though:

1. If we were using the previous block as relay parent, this would imply an additional 6s delay for processing messages, adding latency for users.
2. It gives collators too much freedom.

The problem with (2) is we can not check that a collator actually used the previous relay chain block instead of the most recent one, hence it can cheat and claim block authorship when it was not actually its turn. This works because slots may get occupied by a block, but don't need to, so jumping one slot number is totally fine.

Given that the cheating collator would be the one having the next turn anyways, it would not be able to produce more blocks this way, but can steal a block production opportunity from the previous collator, harming its rewards (potentially benefiting itself) and might also profit from stealing a block with a high amount of transaction fees. A collator might find incentives to harm others and the network itself.

## Proposed Solution

We could use relay chain block numbers as a means for measuring time and thus parachain slots. We would index into the Aura collator set via the relay chain block number the collation is based on (or its parent).

With this we can have the following scheme:

First, we store the relay chain block number that was used for the last produced block in the state of the parachain, we call this block number `s`.

Second when you build a new block we determine the index into the Aura set like this:

We take the block number of the used relay parent `n` determine the index `k` of the legitimate block producer like so:

1. `n >= s + 1 && n < s + 3` -> `k := s + 1`
2. `n <= s` -> illegal
3. `n >= s + 3` -> `k := n - 1`

The idea is, we are logically basing the authorship selection on the parent of the most recent relay chain block: We offset. How the protocol would work for collators:

1. First block produced: Collator takes the most recent relay parent available and uses its parent number `n - 1` to determine whether it is its turn. This number is then stored in the state and will become the current `s := n - 1`.
2. Collators wait 6s (slot time), then they check, either still only the relay parent used by the previous collator is available, then they can use that and `k` will simply be `s+1` or a new relay block already arrived, then they can also use that one (for lower latency), but the `k` will still be determined by `s+1`.

With this we don't need to wait for the next relay chain block to be available, but if it is available we can use it and thus have the lowest possible latency. Because we use `s+1` regardless of whether a newly available relay chain block was used or the previous one, "block stealing" is not possible.

Rule (3) exists to deal with censoring or malfunctioning collators. Without rule (3) we would forever be stuck with the same collator and if it would not produce blocks the chain would stall. With rule (3), it becomes legal to advance `k` by more than `1`, if block production opportunities have been missed.

## Limitations/Consequences

If we take relay chain block numbers to index the collator set, slot times lower than 6s are not possible. I believe that is fine though and might actually lead to simplifications. A slot time of 6s does not mean that we can not produce blocks faster than that, it is just that we can only rotate the block producer every 6s, which does not seem problematic. For elastic scaling, it would simply be the duty of the current collator to produce blocks for all cores of its slot, this is also a very simple solution for how to coordinate what collator should produce blocks for what core ... there is only one -> No coordination required.

## Theoretical Generalization

If we ever found that being able to use the previous relay chain block is not sufficient, the above scheme can be arbitrarily extended at the cost of larger downtime in case of a malfunctioning collator. E.g. if we wanted to allow not only the parent of the most recent block, but even the grand parent, we would achieve that by adjusting above to:

1. `n >= s + 1` && `n < s+ 4` -> `k := s + 1`
2. `n <= s` -> illegal
3. `n >= s + 4` -> `k := n - 1`

## Core Sharing Parachains

We actually do have an issue with parachains sharing a core with this scheme, but that issue exists for the existing time based scheme as well: If paras are sharing a core, e.g. 2 paras, we have assignments like `[A,B,A,B]`, which implies that parachain `A` will only ever see uneven block numbers and `B` will only ever see even block numbers. Thus both paras would operate on half their collator set. For three paras only a third.

This can easily be adjusted statically both in the existing model as in the proposed one. In the time based model you would have a slot time of 12s instead of 6. In the proposed model it is similarly easy to adjust statically for every nth relay chain block.

Ideally though, this would work dynamically, so that chains can switch between core sharings as they please easily. Luckily, we do have a partial solution already in place: The reshuffling of collators every session might be good enough in practice, to not worry about this for now.

## SASSAFRAS

I have not checked, but I would assume that using relay chain block numbers as a measure of time should equally well work here.

## Alternatives

Instead of going with block numbers, we could also keep slots, but make it so, that skipping one slot is illegal, you have to skip two or none. (Same as above proposal with block numbers.)

Then we would store the last slot number or time in the state and do above scheme with slot numbers. (Always advance by 1, regardless of the timestamp of the relay parent as long as it is within bounds.)

In other words, we don't need to move to block numbers to deploy above scheme. It does feel a bit cleaner though, as in the end we are using relay chain blocks as the metric of time anyways.

**Comment by eskimor** [comment]
>  Is it worth to invest time to implement it at this point ?

Very likely not at this point. This just serves as an idea for further improvements, to be picked up when/if needed.

**Comment by burdges** [comment]
Sassafras should slightly improve this:  If you've two relay chain blocks coming then they might come at different times.  We could delay the relay chain aura blocks just slightly, even like 20 miliseconds, so that the praos relay hain blocks tended to arrive first, which then reduces the sassafras advantage.

In test nets, it's possible to reduce the number of praos relay chain blocks, so then you get a measure of how much worse the unpredictable praos blocks make things.  This is much easier to do, even vs delaying down the praos blocks.

**Comment by burdges** [comment]
As an aside, there are academic papers on deducing real time from block time, like https://eprint.iacr.org/2019/838 and https://eprint.iacr.org/2019/1348, but they're after specific threat models, not optimization, so they might be miss-leading here.

**Comment by burdges** [comment]
> Is it worth to invest time to implement it at this point ?

At a high level, we should remember that web2 requires sharded data models everywhere, so elastic scaling tops out quickly once real usage occurs.  We'll risk deluding ourselves into thinking that elastic scaling could go futher than really possible.

It's likely better to push stuff like off-chain messaging rather than squeeking out every last 5% from elastic scaling.

**Comment by sandreim** [comment]
> > Is it worth to invest time to implement it at this point ?
> 
> At a high level, we should remember that web2 requires sharded data models everywhere, so elastic scaling tops out quickly once real usage occurs. We'll risk deluding ourselves into thinking that elastic scaling could go futher than really possible.
> 
> It's likely better to push stuff like off-chain messaging rather than squeeking out every last 5% from elastic scaling.

There are indeed bigger fish to fry, and we have started work on offchain XCMP, but at the same time we want to make Elastic Scaling as robust as possible in order to increase block space utilization.

**Comment by skunert** [comment]
I thought a bit about the problem of older relay chain blocks and think that we can get away with much less.
General Idea:
- Currently we check that para slot roughly matches relay slot
- When building on older parents this is a problem because a collator could just skip some slots and build on a recent parent
- I propose to implement a `ConsensusHook` that requires that blocks fulfill `para_slot = relay_slot + expected_depth`
- From relay chain perspective, we require blocks to be produced in the future
- Since parachain slots can not move backwards and can also not build on a relay chain block of the future, we are guaranteed to always build on an older block

Currently I see no major downsides with this approach and the implementation should be relatively easy.

Drafty Version here: https://github.com/paritytech/polkadot-sdk/pull/7628

cc @eskimor

**Comment by eskimor** [comment]
What is preventing the block producer from just submitting a timestamp of the future, instead of the current one?

**Comment by bkchr** [comment]
Generally I think that @skunert is correct. If you restrict `para_slot == relay_slot` you can do the same calculations are you have done it the first post. So, this means that you "lag behind" with the parachain collator selection, but this would be a constant "lag" and basically just changes which collator is selected.

**Comment by bkchr** [comment]
> So this could be block numbers or relay chain slot numbers like we used before, does not really matter.

Yeah this is what I'm saying. Introducing some new storage item is still very easy to me and would make this implementable very fast. Just checked and we already store the latest `RelayChainSlot`. So, we would not even need any new storage item.
